ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/cw_metrics.py

```diff
@@ -0,0 +1,196 @@
+) -> list[MetricData]:
+    """Generates all of the possible dimensioned (and single undimensioned) metrics from the
+    powerset of the list of dimensions passed-in. Useful as all metrics created by this Lambda
+    have the same value, timestamp, and name and only differ on their aggregations
+
+    Args:
+        metric_name (str): Name of the metric
+        timestamp (datetime): Timestamp to store with the metrics
+        value (float): Value to store with the metrics in each dimension
+        unit (str): The Unit of the metric
+        dimensions (list[dict[str, str]]): The list of dimensions to compute the powerset; this
+        determines the number of metrics that will be stored (2**dimensions.count)
+
+    Returns:
+        list[MetricData]: A list of metrics with each being a set in the powerset of dimensions
+    """
+
+    return [
+        MetricData(
+            metric_name=metric_name,
+            timestamp=timestamp,
+            value=value,
+            # Merge the chain/generator of dimensions of arbitrary size using the "|" operator
+            dimensions=reduce(operator.ior, x, {}),
+            unit=unit,
+        )
+        for x in _powerset(dimensions)
+    ]
+
+
+def put_metric_data(cw_client: Any, metric_namespace: str, metrics: list[MetricData]):
+    """Wraps the boto3 CloudWatch PutMetricData API operation to allow for usage of the MetricData
+    dataclass
+
+    Args:
+        metric_namespace (str): The Namespace of the metric(s) to store in CloudWatch
+        metrics (list[MetricData]): The metrics to store
+    """
+
+    # Convert from a list of the MetricData class to a list of dicts that boto3 understands for this
+    # API. See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Client.put_metric_data
+    metrics_dict_list = [
+        {
+            "MetricName": m.metric_name,
+            "Timestamp": m.timestamp,
+            "Value": m.value,
+            "Unit": m.unit,
+            "Dimensions": [
+                {
+                    "Name": dim_name,
+                    "Value": dim_value,
+                }
+                for dim_name, dim_value in m.dimensions.items()
+            ],
+        }
+        for m in metrics
+    ]
+
+    cw_client.put_metric_data(
+        Namespace=metric_namespace,
+        MetricData=metrics_dict_list,
+    )
+
+
+def get_metric_data(
+    cw_client: Any,
+    metric_data_queries: list[MetricDataQuery],
+    statistic: str,
+    period: int = 60,
+    start_time: datetime = datetime.utcnow() - timedelta(days=15),
+    end_time: datetime = datetime.utcnow(),
+) -> list[MetricDataResult]:
+    """Wraps the GetMetricData CloudWatch Metrics API operation to allow for easier usage. By
+    default, standard resolution metrics from the current time to 15 days in the past are retrieved
+    from CloudWatch Metrics.
+
+    Args:
```

Comment:
Huge nitpick and genuine question - do we need to specify the default values in the docs when it's already specified in the args? Just seems like an extra value to update later if the defaults change

Reply:
(and sorry! I know this is all refactored code and not part of this PR)

Reply:
That's a great point, and honestly I've fallen into the trap of commenting something, changing the behavior of that something, and then forgetting to update the comment to reflect the new behavior more times that I'd like to admit. I think removing this bit from the function's comment is probably best.